zkevm-circuits/src/circuit_tools/cell_manager.rs

```diff
@@ -70,67 +107,389 @@ impl<F: Field> Expr<F> for &Cell<F> {
     }
 }
 
-/// CellType
-#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
-pub enum CellType {
-    /// General
-    Storage,
+#[derive(Clone, Debug)]
+pub struct CellConfig<C: CellTypeTrait> {
+    cell_type: C,
+    num_columns: usize,
+    phase: u8,
+    is_permute: bool,
+}
+
+impl<C: CellTypeTrait> CellConfig<C> {
+    pub fn init_columns<F: Field>(&self, meta: &mut ConstraintSystem<F>) -> Vec<Column<Advice>>{
+        let mut columns = Vec::with_capacity(self.num_columns);
+        match self.phase {
+            0 => columns.iter_mut()
+                    .map(|c| *c = meta.advice_column_in(FirstPhase))
+                    .collect::<Vec<()>>(),
+            1 => columns.iter_mut()
+                    .map(|c| *c = meta.advice_column_in(SecondPhase))
+                    .collect::<Vec<()>>(),
+            2 => columns.iter_mut()
+                    .map(|c| *c = meta.advice_column_in(ThirdPhase))
+                    .collect::<Vec<()>>(),
+            _ => unreachable!(),
+        };
+        if self.is_permute {
+            columns.iter().map(|c| meta.enable_equality(c.clone()));
+        }
+        columns
+    }
 }
 
-/// CellColumn
+
+pub trait CellTypeTrait: 
+    Clone + Copy + Debug + PartialEq + Eq + PartialOrd + Ord + Hash {}
+
+
 #[derive(Clone, Debug)]
-pub struct CellColumn<F> {
-    pub(crate) index: usize,
-    pub(crate) cell_type: CellType,
-    pub(crate) height: usize,
-    pub(crate) expr: Expression<F>,
+pub(crate) struct CellColumn_<F, C> {
+    index: usize,
+    cell_type: C,
+    height: usize,
+    expr: Expression<F>,
+}
+
+
+#[derive(Clone, Debug)]
+pub struct CellManager_<F, C: CellTypeTrait> {
+    max_height: usize,
+    total_width: usize,
+    cells: HashMap<C, (usize, Vec<Cell<F>>)>,
+    cell_configs: Vec<CellConfig<C>>,
+    columns: Vec<CellColumn_<F, C>>,
+}
+
+impl<F: Field, C: CellTypeTrait> CellManager_<F, C> {
+    
+    // 一步到位
+    pub(crate) fn new<T: TableType + Into<C>>(
+        meta: &mut ConstraintSystem<F>,
+        mut cell_configs: Vec<CellConfig<C>>,
+        tables: Vec<&dyn LookupTable_<F, TableType_ = T>>,
+        offset: usize,
+        max_height: usize,
+    ) -> Self {
+
+        let mut cells = HashMap::new();
+        let mut cell_columns = Vec::new();
+        let mut total_width = 0;
+
+        // User need to put fixed columns after advice if have any
+        // permutation is not enable by default
+        for table in tables {
+            if !table.advice_columns().is_empty() {
+                cell_configs.push(CellConfig {
+                    cell_type: table.get_type_().into(),
+                    num_columns: table.advice_columns().len(),
```

Comment:
This loops over the advice/fixed columns and puts it in `cell_configs`, but below these columns are created again in `init_columns` I think? So table columns need to be created twice?

With the super circuit approach, sometimes the table columns will be generated once, and these will have to be reused between different subcircuits that will need to have their own CellManager. So it will need to be possible to do lookups in these premade columns as well, not sure if that is compatible with what is being done here.

Reply:
> This loops over the advice/fixed columns and puts it in cell_configs, but below these columns are created again in init_columns I think? So table columns need to be created twice?

Correct me if I'm wrong :) The `table.advice_columns()` at line 181 are the premade table passed in from ouside, namely the `S` columns in Halo2, but the CM make space for the RAM lookup values at runtime, which is the `A` columns. for So the workflow goes like this:

1. You defined what type of cells you want with `enum CellType` and impl `CellTypeTrait`.
2. Specify configs for each `CellType`, for example: 
```
[
  (ByteLookup, 8 cols, phase1, not permuted), 
  (Storage2, 10 cols, phase2, permuted), 
  ...
]
```
You can basically specify any criteria for any celltypes, skip the celltypes that you don't need, or even repeat the ones with as you wish; the CM is gonna generate them line by line.

3. Pass in `&dyn LookupTable` **as you need**,  CM detects requirements of the tables and append to the list above: 
```
[
  (ByteLookup, 8 cols, phase1, not permuted), 
  (Storage2, 10 cols, phase2, permuted), 
  (Keccack, 5 cols, phase3, not permuted), 
  (Tx, 4 cols, phase3, not permuted)
  ...
]
```
4. At `init_columns` these configs are being carried out as actual advice columns with correct phases and permutation. This function should do the exact same thing as [https://github.com/CeciliaZ030/zkevm-circuits/blob/2c5f3c54f583965cf6efa6d43a8b5baeea3afec6/zkevm-circuits/src/evm_circuit/util.rs#L357-L358](url), so if I'm creating table columns twice then they're probably doing it twice?

Reply:
Ah yeah this makes sense, thanks!